src/index.js

```diff
@@ -4,224 +4,167 @@
+    return this._makePairingChannel(win, channelServerURI, InsecureClientConnection, channelKey);
+  }
 
-  async _writeApplicationData(bytes) {
-    await this._recordSender.withBufferWriter(RECORD_TYPE.APPLICATION_DATA, async buf => {
-      await buf.writeBytes(bytes);
+  static _makePairingChannel(window, wsUri, TlsConnection, psk) {
+    const {WebSocket} = window;
+    const socket = new WebSocket(wsUri);
+    return new Promise((resolve, reject) => {
+      // eslint-disable-next-line prefer-const
+      let stopListening;
+      const onConnectionError = async () => {
+        stopListening();
+        reject(new Error('Error while creating the pairing channel'));
+      };
+      const onFirstMessage = async event => {
+        stopListening();
+        try {
+          const {channelid: channelId} = JSON.parse(event.data);
+          const pskId = utf8ToBytes(channelId);
+          const tlsConnection = await TlsConnection.create(psk, pskId, data => {
+            // To send data over the websocket, it needs to be encoded as a safe string.
+            socket.send(bytesToHex(data));
+          }, window);
+          const instance = new this(channelId, psk, socket, tlsConnection);
+          resolve(instance);
+        } catch (err) {
+          reject(err);
+        }
+      };
+      stopListening = () => {
+        socket.removeEventListener('error', onConnectionError);
+        socket.removeEventListener('message', onFirstMessage);
+      };
+      socket.addEventListener('error', onConnectionError);
+      socket.addEventListener('message', onFirstMessage);
     });
   }
 
-  async _writeHandshakeMessage(msg) {
-    await this._recordSender.withBufferWriter(RECORD_TYPE.HANDSHAKE, async buf => {
-      // XXX TODO: remember buffer position here.
-      await msg.write(buf);
-      // XXX TODO: read back to prev buffer position, add it to the transcript hash.
+  _setupListeners() {
+    this._socket.addEventListener('message', async event => {
+      const channelServerEnvelope = JSON.parse(event.data);
+      const payload = await this._tlsConnection.recv(hexToBytes(channelServerEnvelope.message));
+      if (payload !== null) {
+        const {message, data} = JSON.parse(bytesToUtf8(payload));
+        // Assume the application wants to deal with strings,
+        // not Uint8Array buffers as returned by the TLS layer.
+        this.dispatchEvent({
+          data,
+          message,
+          sender: channelServerEnvelope.sender,
+          type: 'message',
+        });
+      }
     });
+    // Relay the other events.
+    this._socket.addEventListener('error', this.dispatchEvent);
+    this._socket.addEventListener('close', this.dispatchEvent);
   }
 
-  async _flushOutgoingRecord() {
-    const record = await this._recordSender.flush();
-    await this.sendCallback(record);
+  async send(message, data = {}) {
+    const envelope = JSON.stringify({
+      data,
+      message,
+    });
```

Comment:
Our of curiosity, why the "envelope" at this level, rather than letting the caller pass in whatever JSON in whatever envelope format it wants?

Reply:
Also if there's no data this will send `{ "data": {}, "message": "my_message" }` but it might be better to leave out the "data" member entirely in this case?